### PR TITLE
8316233: VirtualThreadStart events should not be thread-filtered

### PR DESCRIPTION
**Pull request by sspitsyn**
The JVMTI VirtualThreadStart events have to follow the ThreadStart events pattern and so, should not be thread-filtered.
The fix includes:
 - `jvmti.xml`: remov the attribute `filtered="thread"` in the `VirtuallThreadStart` event spec
 - `jvmtiEventController.cpp`: remove the `VTHREAD_START_BIT` from the `THREAD_FILTERED_EVENT_BITS` mask and and it to the `NEED_THREAD_LIFE_EVENTS` mask
 - `jvmtiExport.cpp`: rearrangements in the `JvmtiExport::post_vthread_start()` function

The fix also includes a couple of minor unification tweaks:
 - to align `JvmtiExport::post_thread_end()` with `JvmtiExport::post_vthread_end()` to have a unified check for the `JVMTI_PHASE_PRIMORDIAL`.
 -  to rename the local variable `cur_thread` as `thread` to follow the common pattern in `JvmtiExport::post_vthread_start()` and `JvmtiExport::post_vthread_end()`
 
 Testing: ran mach5 tiers 1-6. All tests are passed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316233](https://bugs.openjdk.org/browse/JDK-8316233): VirtualThreadStart events should not be thread-filtered (**Bug** - P4)

### Reviewers
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**) ⚠️ Review applies to [46ce6453](https://git.openjdk.org/jdk/pull/16019/files/46ce64535ab10a9a1e7e2f2c64a5776055238a7f)
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16019/head:pull/16019` \
`$ git checkout pull/16019`

Update a local copy of the PR: \
`$ git checkout pull/16019` \
`$ git pull https://git.openjdk.org/jdk.git pull/16019/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16019`

View PR using the GUI difftool: \
`$ git pr show -t 16019`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16019.diff">https://git.openjdk.org/jdk/pull/16019.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16019#issuecomment-1743906240)

### CODE REVIEW

**Review by lmesnik — CHANGES_REQUESTED** [review]

File: src/hotspot/share/prims/jvmtiExport.cpp [hunk 1]
```diff
@@ -1549,11 +1549,11 @@ void JvmtiExport::post_thread_end(JavaThread *thread) {
 
     JvmtiEnvThreadStateIterator it(state);
     for (JvmtiEnvThreadState* ets = it.first(); ets != nullptr; ets = it.next(ets)) {
+      JvmtiEnv *env = ets->get_env();
```
[the remark this answers is not in the document]
Comment (sspitsyn) [reply src/hotspot/share/prims/jvmtiExport.cpp:1552]: Thanks. I agree with you in general.
The whole fix is relatively small, so I'd prefer to keep such a minor cleanup in this particular case.
I'll update the bug report with this info.

**Review by lmesnik — APPROVED** [review]
Thanks for fixing this.

File: src/hotspot/share/prims/jvmti.xml [hunk 2]
```diff
@@ -13041,7 +13041,7 @@ myInit() {
   </event>
 
  <event label="Virtual Thread Start"
-         id="VirtualThreadStart" const="JVMTI_EVENT_VIRTUAL_THREAD_START" filtered="thread" num="87" phase="start" since="21">
+         id="VirtualThreadStart" const="JVMTI_EVENT_VIRTUAL_THREAD_START" num="87" phase="start" since="21">
```
Comment (plummercj) [inline src/hotspot/share/prims/jvmti.xml:13044]: Does "filtered" mean that the event can be enabled or disabled on a per thread basis, and therefore by removing this it means the event can only be enabled or disabled globally?

Comment (AlanBateman) [reply src/hotspot/share/prims/jvmti.xml:13044]: That's right. The spec for SetEventNotificationMode lists the events cannot be enabled/disabled at the thread level. Both ThreadStart and VirtualThreadStart are listed so I view this JBS/PR issue as fixing the implementation.

File: src/hotspot/share/prims/jvmtiExport.cpp [hunk 3]
```diff
@@ -1575,25 +1575,21 @@ void JvmtiExport::post_vthread_start(jobject vthread) {
...
-  }
-
-  if (state->is_enabled(JVMTI_EVENT_VIRTUAL_THREAD_START)) {
-    JvmtiEnvThreadStateIterator it(state);
+  JavaThread *thread = JavaThread::current();
+  assert(!thread->is_hidden_from_external_view(), "carrier threads can't be hidden");
 
-    for (JvmtiEnvThreadState* ets = it.first(); ets != nullptr; ets = it.next(ets)) {
-      JvmtiEnv *env = ets->get_env();
+  // Do not post virtual thread start event for hidden java thread.
```
[the remark this answers is not in the document]
Comment (sspitsyn) [reply src/hotspot/share/prims/jvmtiExport.cpp:1581]: Good catch, thanks. Forgot to remove this comment. Fixed now.

File: src/hotspot/share/prims/jvmti.xml (hunk 2 above)
Comment (sspitsyn) [reply src/hotspot/share/prims/jvmti.xml:13044]: Thank you for explaining, Alan. There is nothing to add.

File: src/hotspot/share/prims/jvmtiExport.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (plummercj) [reply src/hotspot/share/prims/jvmtiExport.cpp]: I still would like to know how we might end up with a hidden virtual thread.

File: src/hotspot/share/prims/jvmtiExport.cpp [hunk 4]
```diff
@@ -1609,13 +1604,15 @@ void JvmtiExport::post_vthread_end(jobject vthread) {
...
-  JvmtiThreadState *state = get_jvmti_thread_state(cur_thread);
+  JavaThread *thread = JavaThread::current();
+  JvmtiThreadState *state = get_jvmti_thread_state(thread);
   if (state == nullptr) {
     return;
   }
 
-  if (state->is_enabled(JVMTI_EVENT_VIRTUAL_THREAD_END)) {
+  // Do not post virtual thread end event for hidden java thread.
+  if (state->is_enabled(JVMTI_EVENT_VIRTUAL_THREAD_END) &&
```
Comment (alexmenkov) [inline src/hotspot/share/prims/jvmtiExport.cpp:1614]: Should this be assert like in vthread_start?

File: src/hotspot/share/prims/jvmtiExport.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sspitsyn) [reply src/hotspot/share/prims/jvmtiExport.cpp]: A JavaThread can be hidden, not a virtual thread.
For such a case, I'd treat it that a carrier thread is hidden.
The assert is to catch if it ever happens.
Do you think this assert is an overkill?

**Review by alexmenkov — APPROVED** [review]

File: src/hotspot/share/prims/jvmtiExport.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (plummercj) [reply src/hotspot/share/prims/jvmtiExport.cpp]: Never mind. I see you changed it to an assert when your removed the comment.

**Review by plummercj — APPROVED** [review]

File: src/hotspot/share/prims/jvmtiExport.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sspitsyn) [reply src/hotspot/share/prims/jvmtiExport.cpp]: Okay. I agree, it is confusing.

**Comment by sspitsyn** [comment]
Leonid, Alex and Chris, thank you for review!

**Pull request closed without merging**